### PR TITLE
TOML Configuration File

### PR DESCRIPTION
**Pull request by DerLotke**
This PR adds the ability to parse and use configuration files that are stored in TOML-format on the passkey.

The TOML parser is included via the header-only-library `toml.hpp` from https://marzer.github.io/tomlplusplus/index.html and slightly modified in line with its MIT license.

The config module is added to handle loading and distributing configuration information from any part of the application.

### CODE REVIEW

**Comment by DerLotke** [comment]
I forgot to mention: in order to make acquiring the configuration more simple for the user, the behavior of the SDCard class was changed slightly. It is now loaded and unloaded as a singleton (which happens automatically once the config is requested), with the assumption that only one SD card can fit in the passkey device.

File: src/sdcard.cpp [hunk 1]
```diff
@@ -93,3 +95,21 @@ SDCard::SdCardDirectory SDCard::openDir(const String &pathName)
 
     return result;
 }
+
+SDCard& SDCard::load()
+{
+    if (loadedSDCard_ == nullptr)
```
Comment (DGhost001) [inline src/sdcard.cpp]: Let's not do this like that.

Passkey is a multicore µC so, this should be implemented as

```.c
SDCard& SDCard::load()
{
    static SDCard sdCardSingelton;
    return sdCardSingelton;
}
```

or we need a mutex.

And managed pointers are clearly prefered over `new` and `delete`.

File: src/sdcard.cpp [hunk 2]
```diff
@@ -93,3 +95,21 @@ SDCard::SdCardDirectory SDCard::openDir(const String &pathName)
...
+    {
+	loadedSDCard_ = new SDCard;
+    }
+    return *loadedSDCard_;
+}
+
+void SDCard::unload()
+{
```
Comment (DGhost001) [inline src/sdcard.cpp:133]: Is this realy needed? We can't exchange the SD Card without unplugging the power sooooo.....

File: src/keyboard.cpp [hunk 3]
```diff
@@ -35,12 +37,14 @@ UsbKeyboard::UsbKeyboard(bool const skipUsb):
     if(!skipUsb) {
         keyBoard_.begin();
 
+	auto config = getConfig();
```
Comment (DGhost001) [inline src/keyboard.cpp]: Can we get a real type here, please and it looks like `const` is missing.

File: src/config.cpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+#include "sdcard.hpp"
+
+#include <string_view>
+#include <ext/stdio_filebuf.h> 
+#include <iostream>
+#include <fstream>
+#include <string>
+
+static constexpr std::string_view configFile_ = "passkey_config.toml";
+static toml::table * loadedConfig_ = nullptr;
```
Comment (DGhost001) [inline src/config.cpp]: Can we use a managed pointer here ... `shared_ptr` ?

**Review by DGhost001 — CHANGES_REQUESTED** [review]
Needs a bit more work :sunglasses: 

But I like the Idea of a config file stored on the SD Card

File: src/sdcard.cpp (hunk 2 above)
Comment (DerLotke) [reply src/sdcard.cpp:133]: Got removed.

File: src/keyboard.cpp (hunk 3 above)
Comment (DerLotke) [reply src/keyboard.cpp]: Replaced with the actual type that is used (currently a `shared_ptr<toml::table const>`.

File: src/config.cpp (hunk 4 above)
Comment (DerLotke) [reply src/config.cpp]: Uses a shared_ptr now.  Also doesn't keep the configuration file loaded  anymore to safe on DRAM during runtime.

File: src/sdcard.cpp (hunk 1 above)
Comment (DerLotke) [reply src/sdcard.cpp]: I am getting a LoadProhibited error with EXCVADDR 0x00000000 (so apparently a nullpointer exception) when I attempt this. I have opted for an implementation using a shared_ptr for now.

File: src/config.cpp [hunk 5]
```diff
@@ -20,28 +20,22 @@ static toml::table defaultConfig() noexcept
...
-	}
+        std::shared_ptr<SDCard> source = SDCard::load();
+        SDCard::SdCardFile file = source->open(
+    	String(configFileName().data()),
+    	SDCard::OpenMode::FILE_READONLY);
+        if (!file)
+        {
+    	    return std::make_shared<toml::table>(std::move(defaultConfig()));
+        }
+        __gnu_cxx::stdio_filebuf<char> filebuf(file.get(), std::ios::in);
```
Comment (DGhost001) [inline src/config.cpp]: As discussed today... can you add this to the SDCard class so we can hide it there, please?

After that it looks good to me!

**Review by DGhost001 — CHANGES_REQUESTED** [review]
This should be the final cleanup, than we are gooood!

File: src/config.cpp (hunk 5 above)
Comment (DerLotke) [reply src/config.cpp]: Functionality moved to `SDCard::openFileStream`.

**Review by DGhost001 — APPROVED** [review]

**Pull request merged**